### PR TITLE
(feat) O3-4221: Use react hook form and zod in the queue service form

### PR DESCRIPTION
**Pull request by Muppasanipraneeth**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [x] My work includes tests or is validated by existing tests.

### Summary

This PR refactors the existing form validation logic to use [Zod](https://zod.dev/), a TypeScript-first schema validation library. The refactor provides the following benefits: 

- **Type-safe validation**: Ensures better alignment with TypeScript types.  
- **Simplified logic**: Reduces the complexity of form validation.  
- **Improved developer experience**: Allows for inline schema definitions.  
- **Reduced boilerplate**: Removes manual validation code.  

#### Key Changes

1. Introduced a zod schema to define form validation logic.  
2. Replaced manual validation processes with schema-based validation.  
3. Updated test cases to align with the new Zod-based validation approach.  

### Screenshots

See below in PR thread

### Related Issue

- [O3-4221](https://issues.openmrs.org/browse/O3-4221)

### Other Notes

- Minimal impact on the behavior of existing components.  
- Performance improvements by leveraging Zod's efficient validation mechanisms.  

[O3-4221]: https://openmrs.atlassian.net/browse/O3-4221?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by Muppasanipraneeth** [comment]
@denniskigen  as you said made changes

File: packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx [hunk 1]
```diff
@@ -138,31 +153,37 @@ const QueueServiceForm: React.FC<DefaultWorkspaceProps> = ({ closeWorkspace }) =
...
+              render={({ field }) => (
+                <Select
+                  {...field}
+                  id="location"
+                  invalidText="Required"
+                  labelText={t('selectALocation', 'Select a location')}>
+                  {!field.value && <SelectItem text={t('selectALocation', 'Select a location')} value="" />}
+                  {queueLocations.length === 0 && (
+                    <SelectItem text={t('noLocationsAvailable', 'No locations available')} value="" />
+                  )}
```
Comment (denniskigen) [inline packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:168]:
```suggestion
```

I don't think we should do this. Maybe it makes more sense to disable the select input and show a notification below if there are no queue locations.

File: packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx [hunk 2]
```diff
@@ -138,31 +153,37 @@ const QueueServiceForm: React.FC<DefaultWorkspaceProps> = ({ closeWorkspace }) =
...
                 <InlineNotification
                   style={{ margin: '0', minWidth: '100%' }}
                   kind="error"
                   lowContrast
                   title={t('missingLocation', 'Missing location')}
-                  subtitle={t('pleaseSelectLocation', 'Please select a location')}
+                  subtitle={errors.userLocation?.message}
                 />
               </section>
             )}
```
Comment (denniskigen) [inline packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:189]: We should delete this too

File: packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx [hunk 3]
```diff
@@ -138,31 +153,37 @@ const QueueServiceForm: React.FC<DefaultWorkspaceProps> = ({ closeWorkspace }) =
...
-              {queueLocations.length === 0 && (
-                <SelectItem text={t('noLocationsAvailable', 'No locations available')} value="" />
+            <Controller
+              name="userLocation"
+              control={control}
+              render={({ field }) => (
+                <Select
+                  {...field}
+                  id="location"
+                  invalidText="Required"
```
Comment (denniskigen) [inline packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:163]:
```suggestion
								 invalid={errors?.userLocation}
                  invalidText={errors?.userLocation?.message}
```

**Review by denniskigen — COMMENTED** [review]
Thanks for getting this off the ground, @Muppasanipraneeth. I've left some comments.

File: packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx [hunk 4]
```diff
@@ -17,86 +21,91 @@ import { type DefaultWorkspaceProps, restBaseUrl, showSnackbar } from '@openmrs/
...
+          isLowContrast: false,
+          subtitle: error?.message,
+        });
+      });
+  };
 
   return (
-    <Form onSubmit={createQueue} className={styles.form}>
+    <Form onSubmit={handleSubmit(createQueue)} className={styles.form}>
       <Stack gap={4} className={styles.grid}>
```
Comment (denniskigen) [inline packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:84]:
```suggestion
      <Stack gap={5} className={styles.grid}>
```

The extra vertical space looks cleaner to me.

File: packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx [hunk 5]
```diff
@@ -17,86 +21,91 @@ import { type DefaultWorkspaceProps, restBaseUrl, showSnackbar } from '@openmrs/
...
+      queueName: '',
+      queueConcept: '',
+      userLocation: '',
     },
-    [queueName, queueConcept, userLocation, t, closeWorkspace],
-  );
+  });
+
+  const createQueue = (data: QueueServiceFormData) => {
+    saveQueue(data.queueName, data.queueConcept, data.queueName, data.userLocation)
```
[the remark this answers is not in the document]
Comment (Muppasanipraneeth) [reply packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:61]: Screenshot 2024-12-21 at 8 02 40 AM


- I did not changed it is same as before I checked but in this we don't have description  so can pass empty string saveQueue() needs four parameters( queueName: string, queueConcept: string, queueDescription: string, queueLocation: string)

File: packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx [hunk 6]
```diff
@@ -17,86 +21,91 @@ import { type DefaultWorkspaceProps, restBaseUrl, showSnackbar } from '@openmrs/
...
-            isLowContrast: false,
-            subtitle: error?.message,
-          });
-        },
-      );
+  const {
+    control,
+    register,
+    handleSubmit,
+    formState: { errors },
```
[the remark this answers is not in the document]
Comment (Muppasanipraneeth) [reply packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:50]: no link  below I possible provide other wise I figure it out Dennis

[the remark this answers is not in the document]
Comment (Muppasanipraneeth) [reply packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:50]: [https://github.com/openmrs/openmrs-esm-patient-management/blob/aeab2cf252aa506cc8cad1f1c8db96b30909b004/packages/esm-patient-registration-app/src/patient-registration/patient-registration.component.tsx#L190](url)

I am following  like this and the code snippet
Screenshot 2024-12-21 at 11 03 37 AM

[the remark this answers is not in the document]
Comment (denniskigen) [reply packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:50]: https://github.com/openmrs/openmrs-esm-patient-chart/pull/2141

**Comment by Muppasanipraneeth** [comment]
### This is  before  using  Carbon's built-in error handling means while using InlineNotification

https://github.com/user-attachments/assets/fbfda9f1-27ac-4ceb-b252-d79bc771fc78

### This is  after  using  Carbon's built-in error handling 

https://github.com/user-attachments/assets/5b111abf-eea9-4242-968c-46599548893a

File: packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx [hunk 7]
```diff
@@ -1,180 +1,174 @@
...
   TextInput,
+  InlineLoading,
 } from '@carbon/react';
 import { mutate } from 'swr';
 import { type DefaultWorkspaceProps, restBaseUrl, showSnackbar } from '@openmrs/esm-framework';
 import { saveQueue, useServiceConcepts } from './queue-service.resource';
 import { useQueueLocations } from '../create-queue-entry/hooks/useQueueLocations';
 import styles from './queue-service-form.scss';
+import { t } from 'i18next';
+import type { TFunction } from 'i18next';
```
Comment (denniskigen) [inline packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:25]:
```suggestion
```

Import the `TFunction` type from `react-i18next`, and not from `i18next`.

File: packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx [hunk 8]
```diff
@@ -1,180 +1,174 @@
...
+      })
+      .trim()
+      .min(1, t('queueNameRequired', 'Queue name is required')),
+    queueConcept: z.string({
+      required_error: t('queueConceptRequired', 'Queue concept is required'),
+    }),
+    userLocation: z.string({
+      required_error: t('queueLocationRequired', 'Queue location is required'),
+    }),
+  });
```
Comment (denniskigen) [inline packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:41]:
```suggestion
const createQueueServiceSchema = (t: TFunction) =>
  z.object({
    queueName: z
      .string({
        required_error: t('queueNameRequired', 'Queue name is required'),
      })
      .trim()
      .min(1, t('queueNameRequired', 'Queue name is required')),
    queueConcept: z.string({
      required_error: t('queueConceptRequired', 'Queue concept is required'),
    }),
    userLocation: z.string({
      required_error: t('queueLocationRequired', 'Queue location is required'),
    }),
  });
```

File: packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx [hunk 9]
```diff
@@ -1,180 +1,174 @@
...
     },
-    [queueName, queueConcept, userLocation, t, closeWorkspace],
-  );
+  });
+
+  const createQueue = (data: QueueServiceFormData) => {
+    saveQueue(data.queueName, data.queueConcept, '', data.userLocation)
+      .then(() => {
+        showSnackbar({
+          title: t('addQueue', 'Add queue'),
```
Comment (denniskigen) [inline packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:69]:
```suggestion
          title: t('queueServiceCreated', 'Queue service created'),
```

File: packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx [hunk 10]
```diff
@@ -1,180 +1,174 @@
...
+          kind: 'success',
+          subtitle: t('queueAddedSuccessfully', 'Queue added successfully'),
```
Comment (denniskigen) [inline packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:71]:
```suggestion
          subtitle: t('queueServiceCreatedSuccessfully', 'Queue service created successfully'),
```

File: packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx [hunk 11]
```diff
@@ -1,180 +1,174 @@
...
+                <Select
+                  {...field}
+                  labelText={t('selectServiceType', 'Select a service type')}
+                  id="queueConcept"
+                  invalid={!!errors?.queueConcept}
+                  invalidText={errors?.queueConcept?.message}>
+                  {!field.value && <SelectItem text={t('selectServiceType', 'Select a service type')} value="" />}
+                  {queueConcepts.length === 0 && (
+                    <SelectItem text={t('noServicesAvailable', 'No services available')} value="" />
+                  )}
```
Comment (denniskigen) [inline packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:123]:
```suggestion
                  {queueConcepts.length === 0 && (
                    <SelectItem text={t('noServicesAvailable', 'No services available')} value="" />
                  )}
```

I think we can do without this block. Probably better to make the input disabled if there are no services to display and show some helper text below the input.

File: packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx [hunk 12]
```diff
@@ -1,180 +1,174 @@
...
+              name="queueConcept"
+              control={control}
+              render={({ field }) => (
+                <Select
+                  {...field}
+                  labelText={t('selectServiceType', 'Select a service type')}
+                  id="queueConcept"
+                  invalid={!!errors?.queueConcept}
+                  invalidText={errors?.queueConcept?.message}>
+                  {!field.value && <SelectItem text={t('selectServiceType', 'Select a service type')} value="" />}
```
Comment (denniskigen) [inline packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:120]:
```suggestion
                  <SelectItem text={t('selectServiceType', 'Select a service type')} value="" />
```

File: packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx (hunk 8 above)
Comment (Muppasanipraneeth) [reply packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:41]: it looks like same any changes @denniskigen

Comment (Muppasanipraneeth) [reply packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:41]: const createQueueServiceSchema = (t: TFunction) =>
  z.object({
    queueName: z
      .string()
      .trim()
      .nonempty({ message: t('queueNameRequired', 'Queue name is required') }),
    queueConcept: z
      .string()
      .nonempty({ message: t('queueConceptRequired', 'Queue concept is required') }),
    userLocation: z
      .string()
      .nonempty({ message: t('queueLocationRequired', 'Queue location is required') }),
  });

Comment (denniskigen) [reply packages/esm-service-queues-app/src/queue-services/queue-service-form.workspace.tsx:41]: You're right. I've updated the suggestion above.

**Comment by Muppasanipraneeth** [comment]
@denniskigen

**Review by denniskigen — APPROVED** [review]
LGTM. Thanks!

**Comment by Muppasanipraneeth** [comment]
Thanks @denniskigen

**Comment by gracepotma** [comment]
Thank you very much @Muppasanipraneeth for this work - I especially appreciated your detailed PR description and the before vs after videos :D Keep up the good work!

**Comment by denniskigen** [comment]
Can we extend this change to other queue service forms, @Muppasanipraneeth?

**Pull request merged**